design/proposals/variable_provision.md

```diff
@@ -0,0 +1,295 @@
+pre-requisite knowledge a user needs.
+
+An example of this is X.509 Certificates used for authentication, secure communications,
+and data integrity. While it may be simple enough to understand why I would want
+to use Certificates in my applications and infrastructure. Creating and maintaining
+these certificates, and a public key infastructure (PKI) is
+[anything but trivial](https://smallstep.com/blog/everything-pki/). And mistakes
+here can invalidate the security gains from using certificates in the first place.
+
+### Human Intervention Required
+
+As application deployments scale up, the need to automate the creation and rotation
+of credentials will continue to grow as human provisioning of credentails becomes
+a bottleneck. Our API enables end-users to bolt this automation onto Conjur, but
+we have an opportunity to enable more just-in-time creation of these credentails
+on behalf the user to make this kind of automation more widely available.
+
+### Credential Provenance
+
+While the information in Conjur is useful for auditing the *use* of a credential,
+Conjur is missing any ability to audit the *provenance* of a credential or its current
+validity.
+
+## Solution
+
+The solution I propose to address these issues is **variable provisioning**, in which
+Conjur is configured (through the policy where a variable is defined) to provision
+the secret value for a variable when the policy for that variable is loaded.
+
+For example, defining a variable to store an RSA private key could be represented
+in policy as:
+
+```yaml
+- !variable
+  id: private-key
+  annotations:
+    provision/provisioner: rsa
+    provision/rsa/length: 2048
+```
+
+With X.509 in particular, variable provisioning can be use to provide certificate
+signing services managed and trackable entirely within policy. For example, to
+define a root CA in Conjur would be as simple as the following policy. Because
+the properties of the certificate are defined in annotations, it becomes simple
+to inspect the provenance of the certificate value stored in this variable.
+
+```yaml
+- !variable
+  id: private-key
+  annotations:
+    provision/provisioner: rsa
+    provision/rsa/length: 2048
+
+- !variable
+  id: certificate
+  annotations:
+    provision/provisioner: x509-certificate
+    provision/x509-certificate/subject/cn: conjur-master-ca
+    provision/x509-certificate/subject/c: US
+    provision/x509-certificate/basic-constraints/ca: true
+    provision/x509-certificate/basic-constraints/critical: true
+    provision/x509-certificate/key-usage/critical: true
+    provision/x509-certificate/key-usage/key-cert-sign: true
+    provision/x509-certificate/key-usage/crl-sign: true
+    provision/x509-certificate/private-key/variable: ca/private-key
+    provision/x509-certificate/issuer/private-key/variable: ca/private-key # Self-signed CA
+    provision/x509-certificate/issuer/certificate/self: true
+    provision/x509-certificate/ttl: P1Y
+```
```

Comment:
Thinking about IP address whitelisting, I was wondering about using a `Group` as a mechanism for DRYing up some of these annotation.  What do you think about something like:
```yml
- !group 
  id: certificate-base 
  annotations:
    provision/provisioner: x509-certificate
    provision/x509-certificate/subject/c: US
    provision/x509-certificate/basic-constraints/ca: true
    provision/x509-certificate/basic-constraints/critical: true
    provision/x509-certificate/key-usage/critical: true
    provision/x509-certificate/key-usage/key-cert-sign: true
    provision/x509-certificate/key-usage/crl-sign: true
    provision/x509-certificate/private-key/variable: ca/private-key
    provision/x509-certificate/issuer/private-key/variable: ca/private-key # Self-signed CA
    provision/x509-certificate/issuer/certificate/self: true
    provision/x509-certificate/ttl: P1Y

- !variable
  id: certificate
  annotations:
    provision/x509-certificate/subject/cn: conjur-master-ca

- !grant
  member: !variable certificate
  role: !group certificate-base 
```

Reply:
I see where you're going with this, and the issue you're trying to resolve. I'm wondering how you deal with precedence/aggregation when you're a member of multiple groups that define these annotations?

Annotation defaults is more of a `derivativeOf` relationship, rather than the `memberOf` relationship that groups convey.

As a side note, my solution for this was the Policy Factory... to use templates for the defaults, and then create policy instances/resources from the template.

Reply:
Rethinking this, I like your original solution.  It's much more explicit. There are alternatives to streamlining the generation of a large number of resources.